src/dtls_connection_api.c

```diff
@@ -446,159 +519,165 @@ static int dtls_connection_start(void *context_p)
     return ret;
 }
 
-static int dtls_connection_receive(void *context_p, uint8_t *buffer, size_t size,
-                                   void **connection, struct timeval *tv)
+static device_connection_t *dtls_connection_new_incoming(secure_connection_context_t *context,
+                                                         gnutls_dtls_prestate_st *prestate)
 {
-    secure_connection_context_t *context = (secure_connection_context_t *)context_p;
-    int ret, sock, nfds = 0;
-    fd_set read_fds;
     device_connection_t *conn;
-    linked_list_entry_t *conn_entry;
-    const char *err_str;
-    gnutls_dtls_prestate_st prestate;
-    credentials_mode_t ciphersuite;
-
-//  to reduce code redundancy
-    sock = context->conn_listen->sock;
-//  set active file descriptors and calculate required nfds
-    FD_ZERO(&read_fds);
 
-    for (conn_entry = context->connection_list->head; conn_entry != NULL; conn_entry = conn_entry->next)
+    conn = calloc(1, sizeof(device_connection_t));
+    if (conn == NULL)
     {
-        conn = (device_connection_t *)conn_entry->data;
-
-        FD_SET(conn->sock, &read_fds);
-        if (conn->sock >= nfds)
-        {
-            nfds = conn->sock + 1;
-        }
+        return NULL;
     }
 
-    FD_SET(sock, &read_fds);
-    if (sock >= nfds)
-    {
-        nfds = sock + 1;
-    }
+    memcpy(conn, context->conn_listen, sizeof(device_connection_t));
```

Comment:
As far as I understand, things like SSL session and user object (`device_identifier`) should not be copied into a new connection, i.e this should not be a clone operation. In any case "blind" object copy is discouraged and either a dedicated function should be used (which indicates soft or hard cloning), or what's likely more appropriate here is just to set the required fields (sock, addr).